middle_end/flambda/from_lambda/cps_conversion.ml

```diff
@@ -27,40 +27,251 @@ module C = Lambda_conversions
 module Env : sig
   type t
 
-  val create : current_unit_id:Ident.t -> t
+  val create
+     : current_unit_id:Ident.t
+    -> return_continuation:Continuation.t
+    -> exn_continuation:Ilambda.exn_continuation
+    -> t
 
   val current_unit_id : t -> Ident.t
 
-  val add_mutable : t -> Ident.t -> t
   val is_mutable : t -> Ident.t -> bool
   val has_mutables : t -> bool
 
-  (* val add_continuation : t -> int -> t *)
+  val add_mutable_and_make_new_id
+     : t
+    -> Ident.t
+    -> Lambda.value_kind
+    -> t * Ident.t
+
+  val new_id_for_mutable : t -> Ident.t -> t * Ident.t * Lambda.value_kind
+
+  type add_continuation_result = private {
+    body_env : t;
+    handler_env : t;
+    extra_params : (Ident.t * Lambda.value_kind) list;
+  }
+
+  val add_continuation
+     : t
+    -> Continuation.t
+    -> push_to_try_stack:bool
+    -> Asttypes.rec_flag
+    -> add_continuation_result
+
+  val add_static_exn_continuation
+     : t
+    -> int
+    -> Continuation.t
+    -> add_continuation_result
+
+  val get_static_exn_continuation : t -> int -> Continuation.t
+
+  val mark_as_recursive_static_catch : t -> int -> t
+
+  val is_static_exn_recursive : t -> int -> bool
+
+  val get_try_stack : t -> Continuation.t list
+
+  val get_try_stack_at_handler : t -> Continuation.t -> Continuation.t list
+
+  val extra_args_for_continuation : t -> Continuation.t -> Ident.t list
+
+  val extra_args_for_continuation_with_kinds
+     : t
+    -> Continuation.t
+    -> (Ident.t * Lambda.value_kind) list
+
+  val rename_mutable_variable : t -> Ident.t -> Ident.t
 end = struct
   type t = {
     current_unit_id : Ident.t;
-    mutables : Ident.Set.t;
     has_mutables : bool ref; (* for now *)
+    current_values_of_mutables_in_scope
+      : (Ident.t * Lambda.value_kind) Ident.Map.t;
+    mutables_needed_by_continuations : Ident.Set.t Continuation.Map.t;
+    try_stack : Continuation.t list;
+    try_stack_at_handler : (Continuation.t list) Continuation.Map.t;
+    static_exn_continuation : Continuation.t Numbers.Int.Map.t;
+    recursive_static_catches : Numbers.Int.Set.t
   }
 
-  let create ~current_unit_id =
+  let create ~current_unit_id  ~return_continuation
+        ~(exn_continuation : Ilambda.exn_continuation) =
```

Comment:
I'm not sure why you've chosen to pass the exception continuation as the full `Ilambda.exn_continuation` instead of a simple `Continuation.t` (its handler). It's fine either way, but if you're feeling up to a bit of refactoring I would have a slight preference for the `Continuation.t` version.